### ISSUE TITLE
Should HC activation be delayed until needed secrets are available?

### ISSUE THREAD

**Issue by flashyang** [envoyproxy/envoy#17529]
Description:
Recently, we noticed that during the Envoy initialization, there is a race condition between when TLS is configured on a upstream cluster (like validation context) and when active healthcheck begins on that cluster, and it will take around 60s for the Envoy to initialize. In this case, we find that Envoy initiates the first healthcheck on the upstream cluster before the validation context is retrieved, resulting in a health check connection failure and the healthcheck interval will fall back to the no_traffic_interval (because there is no traffic on the cluster). While for Envoy cluster which uses STATIC_DNS and EDS this appears to not delay Envoy initialization, it appears that Envoy cluster using LOGICAL_DNS will wait out the no_traffic_interval to healthcheck again before it considers itself fully initialized.

I have create a same issue https://github.com/envoyproxy/envoy/issues/15977 before, but that issue was closed without a fix. The most close fix is this commit https://github.com/envoyproxy/envoy/pull/16236 but isn't be merged.

**Comment by mattklein123** [comment]
So you are using SDS on the upstream cluster and we are not waiting for SDS to finish before starting health checking? Is that right? If so I agree this should be fixed.

**Comment by flashyang** [comment]
Yes, that's the case. I think we should delay the HC on the upstream cluster until the SDS resources are ready for them.

**Comment by mattklein123** [comment]
OK yes we should fix this. Marking it help wanted. cc @lambdai

**Comment by tnsardesai** [comment]
Hi are there any updates on this ticket? I believe this is still an ongoing issue
